src/librustc_mir_build/hair/pattern/check_match.rs

```diff
@@ -612,81 +617,109 @@ fn check_legality_of_move_bindings(cx: &mut MatchVisitor<'_, '_>, has_guard: boo
     }
 }
 
-/// Check that there are no borrow conflicts in `binding @ subpat` patterns.
+/// Check that there are no borrow or move conflicts in `binding @ subpat` patterns.
 ///
 /// For example, this would reject:
 /// - `ref x @ Some(ref mut y)`,
-/// - `ref mut x @ Some(ref y)`
-/// - `ref mut x @ Some(ref mut y)`.
+/// - `ref mut x @ Some(ref y)`,
+/// - `ref mut x @ Some(ref mut y)`,
+/// - `ref mut? x @ Some(y)`, and
+/// - `x @ Some(ref mut? y)`.
 ///
 /// This analysis is *not* subsumed by NLL.
 fn check_borrow_conflicts_in_at_patterns(cx: &MatchVisitor<'_, '_>, pat: &Pat<'_>) {
-    let tab = cx.tables;
-    let sess = cx.tcx.sess;
-    // Get the mutability of `p` if it's by-ref.
-    let extract_binding_mut = |hir_id, span| match tab.extract_binding_mode(sess, hir_id, span)? {
-        ty::BindByValue(_) => None,
-        ty::BindByReference(m) => Some(m),
+    // Extract `sub` in `binding @ sub`.
+    let (name, sub) = match &pat.kind {
+        hir::PatKind::Binding(.., name, Some(sub)) => (*name, sub),
+        _ => return,
     };
-    pat.walk_always(|pat| {
-        // Extract `sub` in `binding @ sub`.
-        let (name, sub) = match &pat.kind {
-            hir::PatKind::Binding(.., name, Some(sub)) => (*name, sub),
-            _ => return,
-        };
+    let binding_span = pat.span.with_hi(name.span.hi());
 
-        // Extract the mutability.
-        let mut_outer = match extract_binding_mut(pat.hir_id, pat.span) {
-            None => return,
-            Some(m) => m,
-        };
+    let tables = cx.tables;
+    let sess = cx.tcx.sess;
 
-        // We now have `ref $mut_outer binding @ sub` (semantically).
-        // Recurse into each binding in `sub` and find mutability conflicts.
-        let mut conflicts_mut_mut = Vec::new();
-        let mut conflicts_mut_ref = Vec::new();
-        sub.each_binding(|_, hir_id, span, _| {
-            if let Some(mut_inner) = extract_binding_mut(hir_id, span) {
-                match (mut_outer, mut_inner) {
-                    (Mutability::Not, Mutability::Not) => {}
-                    (Mutability::Mut, Mutability::Mut) => conflicts_mut_mut.push(span),
-                    _ => conflicts_mut_ref.push(span),
+    // Get the binding move, extract the mutability if by-ref.
+    let mut_outer = match tables.extract_binding_mode(sess, pat.hir_id, pat.span) {
+        Some(ty::BindByValue(_)) if is_binding_by_move(cx, pat.hir_id, pat.span) => {
+            // We have `x @ pat` where `x` is by-move. Reject all borrows in `pat`.
+            let mut conflicts_ref = Vec::new();
+            sub.each_binding(|_, hir_id, span, _| {
+                match tables.extract_binding_mode(sess, hir_id, span) {
+                    Some(ty::BindByValue(_)) if is_binding_by_move(cx, hir_id, span) => {
+                        sess.delay_span_bug(span, "by-move in subpat unchecked by borrowck");
```

Comment:
Nice catch! I'll add a test to exercise the ICE.

Hmm... I added this mostly as "soundness protection in case borrowck screws up". Should we emit an error here or let the unreachable code be check-pass? From a diagnostics UX perspective it seems not erroring here could be better (to reduce duplicate errors)?

Reply:
I think not erroring here is better.